wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java

```diff
@@ -51,55 +55,66 @@ public class LimitlessRouterServiceImpl implements LimitlessRouterService {
           CACHE_CLEANUP_NANO
       );
 
-  public LimitlessRouterServiceImpl() {
-    this(LimitlessRouterMonitor::new);
+  protected static final SlidingExpirationCacheWithCleanupThread<String, List<HostSpec>>
+      limitlessRouterCache =
+      new SlidingExpirationCacheWithCleanupThread<>(
+          x  -> true,
+          x -> {},
+        CACHE_CLEANUP_NANO
+      );
+
+  public LimitlessRouterServiceImpl(final @NonNull PluginService pluginService) {
+    this(pluginService, LimitlessRouterMonitor::new, new LimitlessQueryHelper(pluginService));
   }
 
-  public LimitlessRouterServiceImpl(final LimitlessRouterMonitorInitializer limitlessRouterMonitorInitializer) {
+  public LimitlessRouterServiceImpl(
+      final @NonNull PluginService pluginService,
+      final LimitlessRouterMonitorInitializer limitlessRouterMonitorInitializer,
+      final LimitlessQueryHelper queryHelper) {
+    this.pluginService = pluginService;
     this.limitlessRouterMonitorInitializer = limitlessRouterMonitorInitializer;
+    this.queryHelper = queryHelper;
   }
 
   @Override
   public List<HostSpec> getLimitlessRouters(final String clusterId, final Properties props) throws SQLException {
     final long cacheExpirationNano = TimeUnit.MILLISECONDS.toNanos(
         MONITOR_DISPOSAL_TIME_MS.getLong(props));
-
-    final LimitlessRouterMonitor
-        limitlessRouterMonitor = limitlessRouterMonitors.get(clusterId, cacheExpirationNano);
-    if (limitlessRouterMonitor == null) {
-      throw new SQLException(
-          Messages.get("LimitlessRouterServiceImpl.nulLimitlessRouterMonitor", new Object[]{clusterId}));
+    final List<HostSpec> limitlessRouters = limitlessRouterCache.get(clusterId, cacheExpirationNano);
+    if (limitlessRouters == null) {
+      return Collections.emptyList();
     }
-    return limitlessRouterMonitor.getLimitlessRouters();
+    return limitlessRouters;
   }
 
   @Override
-  public List<HostSpec> forceGetLimitlessRouters(final String clusterId, final Properties props) throws SQLException {
-
+  public List<HostSpec> forceGetLimitlessRoutersWithConn(
+      final Connection connection, final int hostPort, final Properties props) throws SQLException {
     final long cacheExpirationNano = TimeUnit.MILLISECONDS.toNanos(
         MONITOR_DISPOSAL_TIME_MS.getLong(props));
 
-    final LimitlessRouterMonitor
-        limitlessRouterMonitor = limitlessRouterMonitors.get(clusterId, cacheExpirationNano);
-    if (limitlessRouterMonitor == null) {
-      throw new SQLException(
-          Messages.get("LimitlessRouterServiceImpl.nulLimitlessRouterMonitor", new Object[]{clusterId}));
-    }
     forceGetLimitlessRoutersLock.lock();
```

Comment:
That effectively locks all requests for a router list even these requests are for different clusters.

Reply:
That is not ideal. 
I havent fully fleshed it out, but Im currently looking at Conditions to see if they might work for this. 
https://www.baeldung.com/java-concurrent-locks#working-with-conditions

Reply:
Update: used a map of locks with cluster IDs as keys.